[LLDP] Fix lldpshow script to display multiple MAC addresses on the same remote physical interface

BasimShalata opened the pull request:
<!--
    Please make sure you've read and understood our contributing guidelines:
    https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

    ** Make sure all your commits include a signature generated with `git commit -s` **

    If this is a bug fix, make sure your description includes "closes #xxxx",
    "fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
    issue when the PR is merged.

    If you are adding/modifying/removing any command or utility script, please also
    make sure to add/modify/remove any unit tests from the tests
    directory as appropriate.

    If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
    subcommand, or you are adding a new subcommand, please make sure you also
    update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
    your changes.

    Please provide the following information:
-->

#### What I did

Fix an issue when having the following scenario:
1- remote interface has 2 MACs on the same physical interface.
2- "show lldp table" command displays one entry for only one MAC address

#### How I did it

Root cause:
"show lldp table" command uses lldpshow script to get, parse and display data from the lldp open source package (lldpctl script).
lldpctl script returns a proper info about the 2 MACs but the issue is with the lldpshow script parser where it built a dictionary which its key is the local physical interface.
Therefore when having 2 MACs, lldpctl will return 2 entries but the lldpshow parser will overwrite the first entry.

Fix:
Change the key to be a string of "interface#MAC".
This will enable having 2 entries for 2 different MAC addresses.
   
In addition:
- update display_sum()-->get_summary_output() to return a string instead of printing it directly. This to allow checking the returned value inside the new unit test.

#### How to verify it

- Test with the above scenario and check existence of the 2 MAC addresses in the lldp table.
- Add a new unit test for this scenario.

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)
Output of "show lldp table" remains the same but after the fix will be able to see all MAC addresses learnt on the same physical interface:

# show lldp table
Capability codes: (R) Router, (B) Bridge, (O) Other
LocalPort    RemoteDevice           RemotePortID       Capability    RemotePortDescr
-----------  ---------------------    -----------------     ------------  -----------------
EthernetX                                   xx:xx:xx:xx:xx:xx
EthernetX                                   yy:yy:yy:yy:yy:yy

lguohan commented:
no unit test added?

BasimShalata commented:
I double checked your request and found the following:
1- Currently we don't have any unit test for lldp.
2- "show lldp table" is using the lldpctl script from the lldp package to display lldp entries.
it doesn't use any table from the redis and also we don't have any infrastructure for the lldp unit test to mock the lldpctl.

@lguohan,
do you have any suggestion?

liat-grozovik commented:
@jleveque kindly reminder to assist with pointers to unit test which can be used for this PR.

jleveque commented:
> @jleveque kindly reminder to assist with pointers to unit test which can be used for this PR.

You can mock the lines which call other utilities (like https://github.com/Azure/sonic-utilities/blob/bce123c946a348685675a7f399870cce3acee0da/scripts/lldpshow#L84), and return mocked-up output as if it had been returned by the utility. I can't find an example at the moment, but I'm sure there are some in the existing tests.

BasimShalata commented:
> > @jleveque kindly reminder to assist with pointers to unit test which can be used for this PR.
> 
> You can mock the lines which call other utilities (like
> 
> https://github.com/Azure/sonic-utilities/blob/bce123c946a348685675a7f399870cce3acee0da/scripts/lldpshow#L84
> 
> ), and return mocked-up output as if it had been returned by the utility. I can't find an example at the moment, but I'm sure there are some in the existing tests.

Added a new test which basically doesn't mock the lldpctl as there is more simple way which is:
1- new functionality was added to the parser method and this is the one we need to test.
2- in the unit test, I'm not calling lldp.get_info() which fill lldp.lldpraw DB with an XML output returned from the get_info() which calling lldpctl script, but I'm setting directlly lldp.lldpraw with the info I need to test.
3- in order to test the parser output, I changed the display_sum() method to return a value instead of printing it to the stdout and that's in order to test this returned value.

BasimShalata commented:
/azpw run

liat-grozovik commented:
/azp run

BasimShalata commented:
@lguohan I have added the unit test for this change.
can you please review.

lguohan commented:
@abdosi, can you review this one?

liat-grozovik commented:
@BasimShalata please handle conflicts.
@abdosi please provide your inputs. This fix is required for 202106.

BasimShalata commented:
/azpw run

lguohan reviewed (approved):
:shipit:
